### ISSUE TITLE
Ewelink Error: getaddrinfo ENOTFOUND as-pconnect4.coolkit.cc

### ISSUE THREAD

**Issue by fizulnizam** [bwp91/homebridge-ewelink#224]
### What issue do you have? Please be as thorough and explicit as possible.
My homebridge server will be stopped and when I check the log, it says `Error: getaddrinfo ENOTFOUND as-pconnect4.coolkit.cc
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:67:26) `

### Details of your setup.
* Do you use (1) Homebridge (+ config-ui), (2) Homebridge (CLI) or (3) HOOBS? 
I use HOOBS 

* Which version of Homebridge/HOOBS do you have?
HOOBS 3.3.5

* Which platform do you run Homebridge/HOOBS on (e.g. Raspberry Pi/Windows/HOOBS Box)? Please also mention your version of Node.js/NPM if known.
Raspberry Pi

* Which version of this plugin (homebridge-ewelink) do you have? Has the issue started since upgrading from a previous version?
5.5.1. Yes I also had the same issue on pervious version so I updated it to 5.5.1 hoping it will solve the issue.

* Which eWeLink devices do you have that are causing issues? Please include product models if applicable.
I don’t believe it is specific to one model..

### Please paste any relevant logs below.

```

Error: getaddrinfo ENOTFOUND as-pconnect4.coolkit.cc
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:67:26)

```

**Comment by bwp91** [comment]
Hi, does Homebridge actually stop/shutdown when you get this message?

**Comment by estilles** [comment]
@bwp91 Yes. The Homebridge service stops.

```
3/24/2021, 9:13:36 PM Error: getaddrinfo ENOTFOUND as-pconnect4.coolkit.cc
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
3/24/2021, 9:13:36 PM Got SIGTERM, shutting down Bridge...
```

**Comment by bwp91** [comment]
And when does this appear?

1. When restarting Homebridge?
2. When trying to control a device?
3. When the plugin refreshes the web socket connection every hour?

**Comment by estilles** [comment]
In my particular case, it's been happening at random intervals. I haven't noticed any particular pattern.

FYI ... I just noticed on your `package.json` the plugin required Node `14.16.0`. My HOOBS instance is still on Node 12. Could this be causing the issue?

**Comment by bwp91** [comment]
Won’t have anything to do with the node version.

I have managed to recreate this issue on my own homebridge (by forcing the plugin to use the above address) which makes this a lot easier to debug. 

I’ll get back to you!

**Comment by bwp91** [comment]
@estilles are you able to try the beta I have just pushed, **v5.5.2-6**.
Make sure it's the -6 version.

Some background information:

Once the plugin has logged into eWeLink, they provide a web address for the web socket connection. There are a range of addresses they provide for the Asia server, as-pconnect1.coolkit.cc, as-pconnect2.coolkit.cc, as-pconnect3.coolkit.cc, as-pconnect4.coolkit.cc, etc... The address provided at random and can change each hour that the plugin refreshes the connection.

It seems that the address 'as-pconnect4.coolkit.cc', although provided by eWeLink themselves, is causing these ENOTFOUND errors.

There is no way for the plugin to decide which address is provided by eWeLink, so the fix needs to be a 'workaround'. It seems I need to 'verify' the address provided by ewelink, something I assumed would never need doing...

So in this beta, the plugin will now try and connect to the provided address, and if after 10 seconds it has failed, the plugin will try and request a different address to use. If this 'as-pconnect4.coolkit.cc' is provided again, the same will happen after 10 seconds, the plugin will request a new address. This will loop until a valid address has been given.

I hope this makes sense and the beta should

1. stop homebridge from crashing in this particular 'as-pconnect4.coolkit.cc' case
2. keep requesting a new address until the one provided works correctly

If you can get the beta installed and running for a while, let me know if you continue to get any errors!

**Comment by estilles** [comment]
@bwp91 Thanks for the detailed explanation. I really appreciate it.

I'll install the beta and report back. It's 1am in my time zone, so I'll probably report back sometime tomorrow.

Thank you so much for your help!

**Comment by fizulnizam** [comment]
@bwp91 Thank you so much for the very quick response! However may I ask, how do I install the beta? Can you point me to maybe some documented steps on how do I install the beta? Thank you so much, really appreciate your wonderful work!

**Comment by estilles** [comment]
> @bwp91 Thank you so much for the very quick response! However may I ask, how do I install the beta? Can you point me to maybe some documented steps on how do I install the beta? Thank you so much, really appreciate your wonderful work!

I'm not sure if there's a documented way of doing this, but I manually installed in from the terminal. You can either use the "Terminal" menu option in HOOBS or ssh to `hoobs@hoobs.local` (if you go the ssh route make sure your in the `/home/hoobs/.hoobs/` folder before you run to command below). 

Once you're in the shell you can use the following command
```
npm install homebridge-ewelink@5.5.2-6
```

This is what it looked like for me:
```
hoobs@hoobs:~/.hoobs $ npm install homebridge-ewelink@5.5.2-6

removed 1 package, changed 1 package, and audited 47 packages in 7s

1 package is looking for funding
  run `npm fund` for details

found 0 vulnerabilities
hoobs@hoobs:~/.hoobs $ 
```

Then you can restart the service either from the menu or the terminal. If you choose to do it from the terminal just use the following command:

```
sudo hoobs service restart
```

Hope this helps.

**Comment by bwp91** [comment]
For HOOBS
https://github.com/bwp91/homebridge-ewelink/wiki/Beta-Version#hoobs-ui

Although I am experiencing some strange behaviour with the beta, I will need to look more into it :/

**Comment by bwp91** [comment]
Try beta -7 please!

**Comment by estilles** [comment]
> Try beta -7 please!

Race condition with the `onOpen`  listener?

**Comment by bwp91** [comment]
Looks like there was before! But then changing the ws.open to async meant that the listeners always setup after opening the connection ie the open listener was never called!

**Comment by estilles** [comment]
@bwp91 ... It's a sunny morning here in the Philippines, and I'm happy to report your fix has been holding up! Thank you so much!

**Comment by lingleung2025** [comment]
@bwp91 Thank you for your quick response! I have the same problem here when using hoobs. I will update the bate version 7 tonight(Hong Kong+8). Once I receive any error message, I will let you know^^.

**Comment by bwp91** [comment]
I have just released v5.6.0 (same as beta -7) hopefully (and it looks like) this will sort everything!

**Comment by fizulnizam** [comment]
Thank you so much @bwp91! It solved the issue for me.
